src/Inciter/Diagnostics.C

```diff
@@ -0,0 +1,141 @@
+// *****************************************************************************
+/*!
+  \file      src/Inciter/Diagnostics.C
+  \copyright 2012-2015, J. Bakosi, 2016-2017, Los Alamos National Security, LLC.
+  \brief     Diagnostics class for collecting diagnostics
+  \details   Diagnostics class for collecting diagnostics, e.g., residuals, and
+    various norms of errors while solving partial differential equations.
+*/
+// *****************************************************************************
+
+#include "PDE.h"
+#include "Diagnostics.h"
+#include "DiagReducer.h"
+#include "Discretization.h"
+#include "Inciter/InputDeck/InputDeck.h"
+
+namespace inciter {
+
+extern ctr::InputDeck g_inputdeck;
+extern std::vector< PDE > g_pdes;
+
+static CkReduction::reducerType DiagMerger;
+
+} // inciter::
+
+using inciter::Diagnostics;
+
+Diagnostics::Diagnostics( const Discretization& d )
+// *****************************************************************************
+//  Compute diagnostics, e.g., residuals, norms of errors, etc.
+//! \param[in] d Discretization proxy to read from
+// *****************************************************************************
+{
+  // Store the local IDs of those mesh nodes to which we contribute but do not
+  // own, i.e., slave nodes. Ownership here is defined by having a lower chare
+  // ID than any other chare that also contributes to the node.
+  for (const auto& c : d.Msum())        // for all chares that neighbor our mesh
+    if (d.thisIndex > c.first)          // if our chare ID is larger than theirs
+      for (auto i : c.second)           // store local ID in set
```

Comment:
Not that it matters, but should this be const?

Reply:
You mean variable `i`? If so, it is a copy, so it does not really matter, right?

Reply:
You mean variable i? If so, it is a copy, so it does not really matter, right?

Reply:
Yea, as I said, not important and shouldn't make a difference the compiler, you just always seems very very good at fully qualifying things